go/vt/orchestrator/logic/topology_recovery.go

```diff
@@ -815,113 +613,120 @@ func checkAndRecoverDeadPrimary(analysisEntry inst.ReplicationAnalysis, candidat
 	if !(forceInstanceRecovery || analysisEntry.ClusterDetails.HasAutomatedPrimaryRecovery) {
 		return false, nil, nil
 	}
+	tablet, err := TabletRefresh(analysisEntry.AnalyzedInstanceKey)
```

Comment:
One of the first steps in ERS is `GetTabletMapForShard` which also relies on the topo server. Without this we wont have the list of tablets for promotion. So please correct me if I am wrong, but isn't the topo server already on the hot path of a failover and a failover won't succeed if the topo server is down?

Reply:
In ERS, yes. But I'm comparing it against existent orchestrator. Such dependencies don't exist. I think we should strive to keep those at the bare minimum. The only one that seems unavoidable is locking the shard. Some issues with adding RPC's while performing a failover:
* Overall latency of the failover will increase.
* More likely that it could have transient failures that will leave the shard in a bad state (e.g the call to the topo timeouts due to network hipcups). 

One key difference from this refactor is that people will be relying more and more in ERS to perform failovers.  Before this change that was not the case. For instance, at Slack today we only use ERS for manual / emergency operations. The requirements on those contexts are different than the orchestrator use case. 

What do you think?

Reply:
Note: I think this could be out of scope of this PR, but we need to address this before GA.

Reply:
I think that requiring the topo to be online is required, since it is the place that stores the shard information about the latest primary and the RPC call to promote a replica would fail on the vttablet since it would try to reach the topo server even if we do not try to reach it directly. However, I do think that the number of RPC calls should be minimized, both for better performance and lesser chances of failures.
We want vtorc to eventually reach a state where it uses its ephemeral data to find failover scenarios but only fixes things via code which can be called via RPCs too. This is also the motivation of running `EmergencyReparentShard` from the vtorc and not use a local clone of it which utilises the ephemeral data of vtorc.

Reply:
Dependency on `topo` comes even sooner than this; to run a failover `vtorc` will first run `LockShard`, which uses global topo. The `TabletRefresh` call above comes later. Meaning the dependency on topo is really fundemental and not tied to this or that specific step in the recovery process.
I'm just highlighting this fact.